### DISCUSSION TITLE
"ReferenceError: Request is not defined" when unit testing in Jest

### DISCUSSION THREAD

**Discussion by moonman239** [vercel/next.js#61782] [Help]
### Summary

When I run this code:

```
import { NextRequest } from "next/server";
it("checks for missing token",async ()=>{
    const request = new NextRequest(new Request("https://whatever.app/"));
    
})
```

I get this error: 
"Reference error: Request not defined."

Obviously, I'm doing something "wrong." But what?

**Comment by DevDaniloFerrari** [comment]
Yes, you can mock API calls in Next.js tests using Jest. However, please note that the NextRequest class is specific to the Next.js server environment, and it might not be available in the testing context. Instead, you can use standard Jest mocking techniques to mock the behavior of API calls.

You can use [nock](https://www.npmjs.com/package/nock) to achieve that.

So just run 

`npm install --save-dev nock`

Here is an example:

```
import nock from 'nock';

it("checks for missing token",async ()=>{
  const scope = nock('https://whatever.app/')
    .get('/')
    .once()
    .reply(200, {
      data: 'test data',
    });
})
```

**Reply by moonman239** [reply]
Apologies if I haven't made my intention more clear: I have built an API endpoint in NextJS, and I want to check that it returns the expected response. This endpoint doesn't make any HTTP requests.

Will nock help with this?

**Reply by DevDaniloFerrari** [reply]
Oh I see, in that case you don't need to mock.

Here is an example:

You have an api in `pages/api`

```
import { NextResponse } from 'next/server';

export async function GET() {
  return NextResponse.json({ name: 'Audi' });
}
```

So you can test this way:

```
/**
 * @jest-environment node
 */
import { GET } from "../pages/api/cars/route";

it("check car", async () => {
  const response = await GET();
  const car = await response.json();
  expect(car.name).toBe('Audi');
});
```

Don't forget to put this comment to the top of your test:
```
/**
 * @jest-environment node
 */
```

**Reply by moonman239** [reply]
THanks, but GET requires a NextRequest request.

**Reply by DevDaniloFerrari** [reply]
The error (Reference error: Request not defined.) is still appearing?

**Reply by DevDaniloFerrari** [reply]
If you still facing this error, update your node version.



`nvm install --list`

then

`nvm use 20.11.0` or the version you downloaded.

**Reply by moonman239** [reply]
per your suggestion, I updated Node from 18 to 20.11.0. (I didn't have nvm so I just used the n package)
The reference error still occurs.

**Reply by DevDaniloFerrari** [reply]
Try the command:

`node -v`

and see which version appears.

But I strongly recommend using nvm to manage versions more easily.

If it still doesn't work, try using nvm to change the version and see if it works.

Here it worked with version 20.11.0

![image](https://github.com/vercel/next.js/assets/40414119/f5527e4f-13ea-4222-a424-83d527c6140e)

![image](https://github.com/vercel/next.js/assets/40414119/9bd5e5de-b63c-424d-9473-f0b206b36584)

**Comment by moonman239** [comment]
I think I solved it. Simply create a mock Request object with all the properties relevant to the test.

For example:

```
it("checks for missing token",async ()=>{
    
    const mockRequest = {cookies: {get: (key:string)=>undefined}};
    const response = await GET(mockRequest);
    
})
```

If using TypeScript,  this will generate an error - just ignore it.

**Reply by DevDaniloFerrari** [reply]
I really think that is your node version. But if you solved, ok!
Could you please mark one of my answers as the correct, to close this discussion?

**Comment by icyJoseph** [comment]
@moonman239 in this case the way solutions were recommended was rather noisy, and your solution is also very much not needed. 

Among all of the noise, the answer was given, all you had to do was:

```ts
/**
 * @jest-environment node
 */

import { NextRequest } from "next/server";
it("checks for missing token", async () => {
  const request = new NextRequest(new Request("https://whatever.app/"));
});

```

This is because, in your jest.config.js file, you most likely have:

```js
const customJestConfig = {
  setupFilesAfterEnv: ["<rootDir>/jest.setup.js"],
  testEnvironment: "jsdom", // -> This line, keep it!
};
```

Which tells Jest to use the JSDOM environment, and that one doesn't necessarily contain the Request API. There's a couple of ways to tell Jest which environment to use when testing, but in your case, we can do the one-off magic comment, which sets this test to run in Node, rather than JSDOM.
